zebrad/src/components/mempool.rs

```diff
@@ -261,32 +261,42 @@ impl Service<Request> for Mempool {
                     }
                 }
 
+                // Collect inserted transaction ids.
+                let mut send_to_peers_ids = HashSet::<_>::new();
+
                 // Clean up completed download tasks and add to mempool if successful.
-                // Also, send succesful transactions to peers.
-                let mut inserted_txids = HashSet::new();
                 while let Poll::Ready(Some(r)) = tx_downloads.as_mut().poll_next(cx) {
                     match r {
                         Ok(tx) => {
-                            // Storage handles conflicting transactions or a full mempool internally,
-                            // so just ignore the storage result here
-                            let _ = storage.insert(tx.clone());
-                            // Save transaction ids that we will send to peers
-                            inserted_txids.insert(tx.id);
+                            if let Ok(inserted_id) = storage.insert(tx.clone()) {
+                                // Save transaction ids that we will send to peers
+                                send_to_peers_ids.insert(inserted_id);
+                            }
                         }
                         Err((txid, e)) => {
                             reject_if_needed(storage, txid, e);
                             // TODO: should we also log the result?
                         }
                     };
                 }
-                // Send any newly inserted transactions to peers
-                if !inserted_txids.is_empty() {
-                    let _ = self.transaction_sender.send(inserted_txids)?;
-                }
 
                 // Remove expired transactions from the mempool.
                 if let Some(tip_height) = self.latest_chain_tip.best_tip_height() {
-                    remove_expired_transactions(storage, tip_height);
+                    let expired_transactions = remove_expired_transactions(storage, tip_height);
+
+                    // Remove transactions that are expired from the peers list
+                    let peer_ids_cloned = send_to_peers_ids.clone();
+                    let intersection: HashSet<_> = peer_ids_cloned
+                        .intersection(&expired_transactions)
+                        .collect();
+
+                    send_to_peers_ids.retain(|id| !intersection.contains(id));
+                }
+
+                // Send transactions that were not rejected nor expired to peers
+                let inserted_txids: HashSet<_> = send_to_peers_ids.into_iter().collect();
+                if !inserted_txids.is_empty() {
+                    let _ = self.transaction_sender.send(inserted_txids)?;
                 }
```

Comment:
Code Simplification: We're not changing the `HashSet`'s transaction ID type any more, so we can just send the set we already have.

```suggestion
                // Send transactions that were not rejected nor expired to peers
                if !send_to_peers_ids.is_empty() {
                    let _ = self.transaction_sender.send(send_to_peers_ids)?;
                }
```

Reply:
Ops, thanks. Fixed as part of https://github.com/ZcashFoundation/zebra/pull/2858/commits/16f8ea89f2d629fa2e16248cfcbb59e1c29112dc